explain/agents.py

```diff
@@ -0,0 +1,136 @@
+"""
+Agent registry system for AI coding assistants.
+"""
+
+from __future__ import annotations
+
+import os
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, ClassVar
+
+
+@dataclass
+class BaseAgent(ABC):
+    """Base interface for AI coding assistants."""
+
+    agent_bin: Path
+    log_level: str = "CRITICAL"
+
+    name: ClassVar[str]
+    program_name: ClassVar[str]
+    display_name: ClassVar[str]
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        # Subclasses must set up their identifying strings.
+        required = ["name", "program_name", "display_name"]
+        if not all(hasattr(cls, f) for f in required):
+            raise ValueError(
+                f"Agent class {cls.__name__} must define non-empty attributes: {','.join(required)}"
+            )
+
+        AgentRegistry._agents[cls.name] = cls
+
+    @classmethod
+    @abstractmethod
+    def find_binary(cls) -> Path | None:
```

Comment:
This could work as well:
```
import abc
import functools


class Base(abc.ABC):
    @property
    @abc.abstractmethod
    def binary(self) -> Path | None: ...


class Deriv(Base):
    @functools.cached_property
    def binary(self) -> Path | None:
        print("called")
        return Path("/foo")


d = Deriv()
print(d.binary)
print(d.binary)
```

Reply:
I think it's fine either way!